### PR TITLE
add structured planning agent

### PR DESCRIPTION
**Pull request by logan-markewich**
This PR adds a new `AgentRunner` that works with any agent worker to 
- create an initial plan to accomplish a given task
- can run steps concurrently when they share no dependencies
- refines the plan after each step

There is also a very comprehensive notebook added that details how it works for both high-level and low-level APIs.

### CODE REVIEW

**Comment by logan-markewich** [comment]
TODO: add some unit tests for sanity

File: llama-index-core/llama_index/core/agent/runner/planner.py [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+    def __init__(
+        self,
+        agent_worker: BaseAgentWorker,
+        tools: Optional[List[BaseTool]] = None,
+        tool_retriever: Optional[ObjectRetriever[BaseTool]] = None,
+        chat_history: Optional[List[ChatMessage]] = None,
+        state: Optional[PlannerAgentState] = None,
+        memory: Optional[BaseMemory] = None,
+        llm: Optional[LLM] = None,
+        initial_plan_prompt: str = DEFAULT_INITIAL_PLAN_PROMPT,
```
Comment (jerryjliu) [inline llama-index-core/llama_index/core/agent/runner/planner.py:166]: nit: most of our existing fn signatures take in Optional[PromptTemplate] as input. If you still want to allow strings could do Optional[Union[PromptTemplate, str]]

**Review by jerryjliu — COMMENTED** [review]
This is cool. I like that structured planning on the agent runner allows us to do orchestration optimizations. Some high-level comments: 

- [future] Would be good to communicate in the docs when developers should use a structured planning agent runner vs. a regular agent runner. I see two situations:
   - they're deciding what agent runner to use with a prepackaged agent worker, like ReAct. maybe they don't know/don't care and just want something to work
   - they're trying to build a custom agent. what logic should they put in the agent worker, and how does using structured planning vs. regular agent runner affect that. i'm assuming they're not meant to customize the agent runner? 


- you prob already know this, but implementing LATS doesn't really fit into this structured planning abstraction. any plan generated in the structured runner is a deterministic DAG. but in LATS it's really about planning out an entire distribution of possible futures from the given state (instead of just one in the deterministic case), and figuring out which future we should rollout each stage. wonder if an MCTS agent runner would eventually make sense

File: llama-index-core/llama_index/core/agent/runner/planner.py (hunk 1 above)
Comment (logan-markewich) [reply llama-index-core/llama_index/core/agent/runner/planner.py]: Done!

**Comment by logan-markewich** [comment]
I think in general, users could definitely customize the agent runner. In particular, the `create_tasks` and `refine_plan` functions could be overridden to implement custom planning logic. 

That might make a fun "from scratch" notebook later, something that plans a simple list of tasks (would be easier for open-source LLMs to do)

**Pull request merged**